File: kraken/src/main/java/glitch/kraken/services/ChatService.java
package glitch.kraken.services;

import glitch.api.http.Routes;
import glitch.api.http.Unofficial;
import glitch.api.objects.json.interfaces.OrdinalList;
import glitch.kraken.GlitchKraken;
import glitch.kraken.object.json.*;
import glitch.kraken.object.json.collections.ChatRooms;
import glitch.kraken.object.json.collections.EmoteSets;
import glitch.kraken.object.json.collections.Emotes;
import glitch.service.AbstractHttpService;
import java.util.Arrays;
import java.util.Collection;
import java.util.stream.Collectors;
import reactor.core.publisher.Flux;
import reactor.core.publisher.Mono;

public class ChatService extends AbstractHttpService {
    public ChatService(GlitchKraken rest) {
        super(rest.getClient(), rest.getHttpClient());
    }

    public Mono<ChatBadges> getChatBadges(Long id) {
        return exchangeTo(Routes.get("/chat/%s/badges").newRequest(id), ChatBadges.class);
    }

    public Mono<EmoteSets> getChatEmoteSets(Integer... ids) {
        return getChatEmoteSets(Arrays.asList(ids));
    }

    public Mono<EmoteSets> getChatEmoteSets(Collection<Integer> ids) {
        return exchangeTo(Routes.get("/chat/emoticon_images").newRequest()
                .queryParam("emotesets", ids.stream()
                        .map(String::valueOf)
                        .collect(Collectors.joining(","))), EmoteSets.class);
    }

    public Flux<Emote> getAllEmoticons() {
        return exchangeTo(Routes.get("/chat/emoticons").newRequest(), Emotes.class)
                .flatMapIterable(OrdinalList::getData);
    }

    public Flux<ChatRoom> getChatRooms(Long id) {
        return exchangeTo(Routes.get("/chat/%s/rooms").newRequest(id), ChatRooms.class)
                .flatMapIterable(OrdinalList::getData);
    }

    @Unofficial
    public Mono<GlobalUserState> getChatUserState(User user) {
        return getChatUserState(user.getId());
    }

    @Unofficial
    public Mono<GlobalUserState> getChatUserState(Long id) {
        return exchangeTo(Routes.get("/users/%s/chat").newRequest(id), GlobalUserState.class);
    }
}

File: kraken/src/main/kotlin/io/glitchlib/v5/service/ChatService.kt
package io.glitchlib.v5.service

import io.glitchlib.GlitchClient
import io.glitchlib.Unofficial
import io.glitchlib.internal.http.bodyFlowable
import io.glitchlib.internal.http.bodySingle
import io.glitchlib.model.OrdinalList
import io.glitchlib.v5.internal.AbstractKrakenService
import io.glitchlib.v5.model.json.Channel
import io.glitchlib.v5.model.json.ChatBadges
import io.glitchlib.v5.model.json.Emote
import io.glitchlib.v5.model.json.EmoteSets
import io.glitchlib.v5.model.json.User

class ChatService(client: GlitchClient) : AbstractKrakenService(client) {

    val allEmoticons
        get() = get<OrdinalList<Emote>>("/chat/emoticons").bodyFlowable

    fun getChatBadges(id: Long) =
            get<ChatBadges>("/chat/$id/badges").bodySingle

    fun getChatEmoteSets(vararg ids: Int) = getChatEmoteSets(ids.toList())

    fun getChatEmoteSets(ids: Collection<Int>) =
            get<EmoteSets>("/chat/emoticon_images") {
                addQueryParameters("emotesets", ids.joinToString(",") { it.toString() })
            }.bodySingle

    fun getChatRooms(id: Long) = client.getChatRooms(id)

    fun getChatRooms(user: Channel) = client.getChatRooms(user.id)

    @Unofficial
    fun getChatUserState(user: User) = getChatUserState(user.id)

    @Unofficial
    fun getChatUserState(id: Long) = client.getChatUser(id)
}
